src/rafimage.cpp

```diff
@@ -289,17 +291,31 @@ namespace Exiv2 {
 
         clearMetadata();
 
-        io_->seek(84,BasicIo::beg);
+        if (io_->seek(84,BasicIo::beg) != 0) throw Error(kerFailedToReadImageData);
         byte jpg_img_offset [4];
-        io_->read(jpg_img_offset, 4);
+        if (io_->read(jpg_img_offset, 4) != 4) throw Error(kerFailedToReadImageData);
         byte jpg_img_length [4];
-        io_->read(jpg_img_length, 4);
-        long jpg_img_off = Exiv2::getULong((const byte *) jpg_img_offset, bigEndian);
-        long jpg_img_len = Exiv2::getULong((const byte *) jpg_img_length, bigEndian);
+        if (io_->read(jpg_img_length, 4) != 4) throw Error(kerFailedToReadImageData);
+        uint32_t jpg_img_off_u32 = Exiv2::getULong((const byte *) jpg_img_offset, bigEndian);
+        uint32_t jpg_img_len_u32 = Exiv2::getULong((const byte *) jpg_img_length, bigEndian);
+
+        enforce(Safe::add(jpg_img_off_u32, jpg_img_len_u32) <= io_->size(), kerCorruptedMetadata);
```

Comment:
Out of interest: where did you take this check from? It makes sense to me, checking that offset+length are not out of bounds of the image, but I am not really familiar with the intricacies of this image format.

Reply:
I deduced it from the code below. It seeks `jpg_img_off + 12` from the beginning of the file, then reads `jpg_img_len - 12` bytes. I am also pretty sure there was a bug in the `io_->read` on line 302, because it used to read `buf.size_ - 12` bytes, but then passes `buf.size_` bytes to `TiffParser::decode`. So I changed that line of code to read `buf.size_` bytes, which I am pretty sure is what is what it ought to do.